app/lib/authorizer.rb

```diff
@@ -0,0 +1,144 @@
+#
+# This class should be fully tested, and all changes should be reviewed.
+#
+# Controller usage:
+#   student = authorized { Student.find(params[:id]) }
+#   homeroom = authorized { Homeroom.(params[:id]) }
+#   students = authorized { homeroom.students }
+#
+# Standalone usage:
+#   authorizer = Authorizer.new(educator)
+#   student = authorizer.authorized { Student.find(params[:id]) }
+#   homeroom = authorizer.authorized { Homeroom.(params[:id]) }
+#   students = authorizer.authorized { homeroom.students }
+#
+# Instead of:
+#   Homeroom.find(id).students # no authorization check
+#   Section.find(id).students # no authorization check
+#   current_educator.homeroom.students # doesn't work for all roles because
+#     some roles don't have homeroom associations (e.g. if they're school-wide)
+#   current_educator.school.students # doesn't work for all roles because
+#     some roles don't have school associations (e.g. if they're district-wide)
+
+#
+class Authorizer
+  # These fields are required on the `Student` model for
+  # authorization checks
+  def self.student_fields_for_authorization
+    [
+      :id,
+      :program_assigned,
+      :limited_english_proficiency,
+      :school_id,
+      :grade
+    ]
+  end
+
+  def initialize(educator)
+    @educator = educator
+    @authorized_dispatcher = AuthorizedDispatcher.new(self)
+  end
+
+  # Usage:
+  #  student = authorized { Student.find(params[:id]) }
+  def authorized(&block)
+    @authorized_dispatcher.dispatch(&block)
+  end
+
+  # This method is the source of truth for whether an educator is authorized to view information about a particular
+  # student.
+  def is_authorized_for_student?(student)
+    begin
+      return true if @educator.districtwide_access?
+
+      return false if @educator.restricted_to_sped_students && !(student.program_assigned.in? ['Sp Ed', 'SEIP'])
+      return false if @educator.restricted_to_english_language_learners && student.limited_english_proficiency == 'Fluent'
+      return false if @educator.school_id.present? && student.school_id.present? && @educator.school_id != student.school_id
+
+      return true if @educator.schoolwide_access? || @educator.admin? # Schoolwide admin
+      return true if @educator.has_access_to_grade_levels? && student.grade.in?(@educator.grade_level_access) # Grade level access
+      return true if student.in?(@educator.students) # Homeroom level access
+      return true if student.in?(@educator.section_students) # Section level access
+    rescue ActiveModel::MissingAttributeError => err
+      # We can't do authorization checks on models with `#select` that are missing
+      # fields.  If this happens, it's probably because the developer is trying to
+      # to optimize a query.  The authorization layer could only recover by making more
+      # queries, so instead we raise and force the developer to figure out how to resolve.
+      #
+      # See `Authorizer.student_fields_for_authorization`
+      raise err
```

Comment:
I have a question about the raise-and-rescue pattern here. We're not doing any logging or any side effect creation with the error message, just re-raising. And we're not in `map` or `each` loop. So is the intent of this raise-and-rescue mainly to catch the eye of developers who might raise this error, and provide a context for the code comments? Or is it meant to do something else that I'm not understanding yet?

Reply:
Yeah - this problem here (with using `select`) is an edge case, but it came up in the `lasids` endpoint, which is how I realized it.

So yeah, I was thinking about catching the eye of developers, so added in a comment, and added in the `begin/raise` to make explicit what's happening (we could also add a log message in the raise, I think that's what I did initially in trying to understand what was happening).

If this seems weird, we could remove it for just comments, or we could improve the comments to make it clearer why it's like this or what's happening.

Reply:
OK, I see! So right now the `lasids` endpoint looks like this:

```
def lasids
  render json: Student.pluck(:local_id)
end
```

There's no explicit authorization check, and that's because authorization is handled at the controller level like this:

```
before_action :authorize_for_districtwide_access_admin, only: [:lasids]
```

So basically, part of this PR is moving us to a more standardized way of handling things here, where we'd want to call `authorized` in our `lasids` endpoint code, even though there's controller authorization, because that's a more fail-proof and centralized way of doing things. Does that sound about right? 

And as we keep implementing this work, we'd move from `:pluck` (which only works when we assume the educator is a districtwide admin and can access everything) to some kind of `select` statement.